crates/bevy_core_pipeline/src/bloom/bloom.wgsl

```diff
@@ -1,138 +1,137 @@
+// Bloom works by creating an intermediate texture with a bunch of mip levels, each half the size of the previous.
+// You then downsample each mip (starting with the original texture) to the lower resolution mip under it, going in order.
+// You then upsample each mip (starting from the smallest mip) and blend with the higher resolution mip above it (ending on the original texture).
+//
+// References:
+// * http://www.iryoku.com/next-generation-post-processing-in-call-of-duty-advanced-warfare
+// * https://learnopengl.com/Guest-Articles/2022/Phys.-Based-Bloom
+
 #import bevy_core_pipeline::fullscreen_vertex_shader
+#import bevy_core_pipeline::tonemapping
 
-struct BloomUniforms {
-    threshold: f32,
-    knee: f32,
-    scale: f32,
-    intensity: f32,
+struct BloomDownsamplingUniforms {
+    threshold_precomputations: vec4<f32>,
     viewport: vec4<f32>,
 };
 
 @group(0) @binding(0)
-var original: texture_2d<f32>;
+var input_texture: texture_2d<f32>;
 @group(0) @binding(1)
-var original_sampler: sampler;
-@group(0) @binding(2)
-var<uniform> uniforms: BloomUniforms;
-@group(0) @binding(3)
-var up: texture_2d<f32>;
+var s: sampler;
 
-fn quadratic_threshold(color: vec4<f32>, threshold: f32, curve: vec3<f32>) -> vec4<f32> {
-    let br = max(max(color.r, color.g), color.b);
-
-    var rq: f32 = clamp(br - curve.x, 0.0, curve.y);
-    rq = curve.z * rq * rq;
+#ifdef FIRST_DOWNSAMPLE
+@group(0) @binding(2)
+var<uniform> uniforms: BloomDownsamplingUniforms;
+
+// https://catlikecoding.com/unity/tutorials/advanced-rendering/bloom/#3.4
+fn soft_threshold(color: vec3<f32>) -> vec3<f32> {
+    let brightness = max(color.r, max(color.g, color.b));
+    var softness = brightness - uniforms.threshold_precomputations.y;
+    softness = clamp(softness, 0.0, uniforms.threshold_precomputations.z);
+    softness = softness * softness * uniforms.threshold_precomputations.w;
+    var contribution = max(brightness - uniforms.threshold_precomputations.x, softness);
+    contribution /= max(brightness, 0.00001); // Prevent division by 0
+    return color * contribution;
+}
+#endif
 
-    return color * max(rq, br - threshold) / max(br, 0.0001);
+// http://graphicrants.blogspot.com/2013/12/tone-mapping.html
+fn karis_average(color: vec3<f32>) -> f32 {
+    let luma = tonemapping_luminance(rgb_to_srgb_simple(color)) / 4.0;
+    return 1.0 / (1.0 + luma);
```

Comment:
I think this function is wrong. The division by 4 is, as far as I can tell, to average the luminance over 4 samples, but the colour pass in is the filter-weighted average of 4 samples in each group. So the returned weight seems to be incorrect.

Reply:
I still think this is wrong though. I don't see why would be divide the luminance by 4 here given the samples were summed and averaged already.

Reply:
As I say in the other reply, I think this is just magic. The overall image is somehow not affected if you bypass this.

Reply:
The learnopengl article says the * 0.25 is because the value passed in is the sum of 4 samples. But there is a / 4 applied to each group. Hmmmm

Reply:
Clarity, as this matches the learnopengl reference material, until we know this is wrong, I'm now happy to consider this resolved, pending a response from Jimenez.